src/init/VarResolver.ts

```diff
@@ -0,0 +1,139 @@
+import yargs from 'yargs';
+import { AsyncHandler } from '../util/handlers/AsyncHandler';
+import { ensureTrailingSlash, modulePathPlaceholder, resolveAssetPath } from '../util/PathUtil';
+
+const defaultConfig = `${modulePathPlaceholder}config/default.json`;
+const defaultVarConfig = `${modulePathPlaceholder}config/app-setup/vars.json`;
+
+/**
+ * CLI options needed for performing meta-process of app initialization.
+ * These options doesn't contribute to components-js vars normally.
+ */
+export const baseYargsOptions: Record<string, yargs.Options> = {
+  config: { type: 'string', alias: 'c', default: defaultConfig, requiresArg: true },
+  loggingLevel: { type: 'string', alias: 'l', default: 'info', requiresArg: true },
+  mainModulePath: { type: 'string', alias: 'm', requiresArg: true },
+  varConfig: { type: 'string', alias: 'v', default: defaultVarConfig, requiresArg: true },
+};
+
+/**
+ * A handler that takes args, and returns computed variable value
+ */
+export abstract class VarComputer extends AsyncHandler<yargs.Arguments, any> {
+  public abstract handle(args: yargs.Arguments): Promise<any>;
+}
+
+// Currently componentsjs-builder giving error, if we use external TSTypeReference.
+// Re-definition is temporary work around.
+export interface IYargsOptions extends yargs.Options {
+  alias?: string | readonly string[];
+  default?: any;
+  desc?: string;
+  requiresArg?: boolean;
+  type?: 'array' | 'count' | yargs.PositionalOptionsType;
+}
+
+export interface IVarResolverOptions {
+  yargsOptions: Record<string, IYargsOptions>;
+  varComputers: Record<string, VarComputer>;
+  usage?: string;
+  strictMode?: boolean;
+}
+
+export class VarResolver {
+  protected readonly opts: IVarResolverOptions;
+  protected readonly effectiveYargsOptions: Record<string, yargs.Options>;
+
+  public constructor(opts: IVarResolverOptions) {
```

Comment:
Json range doesn't do type-check. Hence interface is used. should we remove it then?

Reply:
That's a fair point, we do lose the type check in components.js. But I think this is acceptable here since it will make the var config a lot more readable for users. In case the input are wrong the server is also immediately going to error so there's no risk of starting the server with a faulty config.